app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt

```diff
@@ -81,82 +85,208 @@ object FirebaseDB {
+
+        metadata.pictureId = String.format("%s_%s", userId, "profile_picture")
+
+        return storePicture(image, USER_PROFILE_PICTURE_METADATA, metadata,
+            databaseReference.child(USERS_PATH).child(userId),
+            storageReference.child(USERS_PATH).child(userId).child(
+                USER_PROFILE_PICTURE_METADATA))
+    }
+
     /**
      * Adds and [image] for the user [userId] in the database
      *
      * @param image the image we want to add to the database
      * @param userId the ID of the user wanting to store a new image
      * @return a unique ID under which the image got stored, and null if the image couldn't get stored
      */
-    suspend fun addImage(image: Bitmap, userId: Int): String? {
+    suspend fun addImage(image: Bitmap, userId: Int, metadata: PhotoMetadata): PhotoMetadata? {
 
-        // Points to the node where we want to store the metadata of the image in the realtime database
-        val imageMetadataNode = databaseReference.child("pictures").child("userId")
-            .child(userId.toString())
-        // Generates a new imageID for the image we're going to add, and return null if it fails
-        val uniqueID = imageMetadataNode.push().key ?: return null
+        return storePicture(image, null, metadata,
+            databaseReference.child("pictures").child(userId.toString()),
+            storageReference.child("userId").child(userId.toString()))
+    }
 
-        // Points to the node where we want to store the image in Firebase storage
-        val imageNode = storageReference.child("userId").child(userId.toString()).child(uniqueID)
-        // Compress the image in a JPEG format
-        val compressedImage = ByteArrayOutputStream()
-        image.compress(Bitmap.CompressFormat.JPEG, 100, compressedImage)
+    /**
+     * Test whether a [User] exists in the database for the given [userId]
+     *
+     * @param userId The id of the user to check for the existence in the database
+     * @return true if the user exists, false otherwise
+     */
+    suspend fun userExists(userId: String): Boolean {
         return try {
-            // Upload the image
-            imageNode.putBytes(compressedImage.toByteArray()).await()
-            // Upload the image's metadata (might need to add/remove fields later, and pass them as args)
-            val imageMetadataMap = HashMap<String, Any>()
-            imageMetadataMap["time"] = System.currentTimeMillis()
-            imageMetadataMap["type"] = "plastic bottle"
-
-            imageMetadataNode.child(uniqueID).setValue(imageMetadataMap)
-            uniqueID
-        } catch (e: Error) {
-            null
+            withTimeout(TIMEOUT) {
+                if (userId.isBlank())
+                    throw java.lang.IllegalArgumentException("The userId cannot be a blank string")
+
+                val data = databaseReference.child(USERS_PATH).child(userId).get().await()
+                data.exists()

```

Comment:
That's true, by going one level deeper and checking the `id` node of the `User` we will use less resources as we won't take the whole `User` but only its `id` that a great performance improvement thanks!

However, `data != null` doesn't work to test if the data under the particular node exists or not, because `data` is a `DataSnapshot`, so even if there is no valid data under this snapshot, the variable `data` is not `null`. So we have to explicitly call the `.exists()` method.

Reply:
Of course, sorry for this mistake!